ip2asn.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python3
+"""Takes ip addresses provided on stdin (one per line) and outputs info about the ASNs that contain them."""
+#Appears to work just fine under both python2 and python3.
+
+#Download https://iptoasn.com/data/ip2asn-combined.tsv.gz to current directory, gunzip it.
+
+import os
+import sys
+import csv
+import fileinput
+from ipaddress import summarize_address_range, IPv4Address, IPv6Address		#Not needed: ip_address, AddressValueError
+
+
+#Note: netaddr removed as standard python library does not include it, but appears to include ipaddress.  See v0.1 for comparison between the two - both provide equal results for all ip ranges.
+ip2asn_version = '0.6.1'
+
+
+def load_asn_table(source_file):
+	"""Loads the subnets from ip2asn-combined.tsv."""
+	#real	0m46.232s to load
+
+	Paranoid = False			#Setting this to True enables additional checks on the ip2asn-combined raw data (that descriptions and countries stay consistent)
+
+	as_info_struct = {}		#Dictionary of Lists of dictionaries.
+					#Top level dictionary has keys = 4_firstoctet for ipv4, 6_firsthexbyte for ipv6.  Values are the lists on the next line
+					#next level lists have keys 0-128; their values are dictionaries.  For most specific subnet, search from 128 back to 1 (32 back to 1 for ipv4)
+					#second level dictionaries;  key = IP object, value is as_num
+					#Adding the first_octet level makes a significant and visual performance increase in lookup time.
+
+	asn_country = {}		#Key= as_num, value=2 letter country code
+	asn_description = {}		#Key= as_num, value=asn description
+
+
+	if os.path.exists(source_file):
+		with open(source_file, 'r') as aih:
+			reader = csv.reader(aih, delimiter='\t')
+			#Format: range_start range_end AS_number country_code AS_description
+			for first_ip, last_ip, as_num, country, as_description in reader:
+				if sys.version_info < (3, 0):
```

Comment:
Consider using `six` to implement python version compatibility.
https://stackoverflow.com/a/29213992